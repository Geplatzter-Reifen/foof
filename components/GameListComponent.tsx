import { BoardGame } from "@/model/model";
import { View, Text } from "react-native";

export default function GameListComponent({ games }: { games: BoardGame[] }) {
  return (
    <View>
      {games.map((game) => (
        <View key={game.title}>
          <Text>{game.title} - </Text>
          <Text>Min Players: {game.minPlayers}</Text>
        </View>
      ))}
    </View>
  );
}
