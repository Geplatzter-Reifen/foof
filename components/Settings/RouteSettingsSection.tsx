import React, { useEffect, useState } from "react";
import {
  Icon,
  IconElement,
  Text,
  ThemeType,
  useTheme,
} from "@ui-kitten/components";
import {
  ImageProps,
  StyleSheet,
  Alert,
  View,
  Linking,
  TouchableOpacity,
} from "react-native";
import { useLocalSearchParams } from "expo-router";
import {
  deleteRoute,
  getTourRoute,
  setTourRoute,
} from "@/services/data/routeService";
import * as DocumentPicker from "expo-document-picker";
import * as FileSystem from "expo-file-system";
import * as Gjv from "geojson-validation";
import SingleSettingLayout from "@/components/Settings/SingleSettingLayout";
import InlineRow from "@/components/Settings/InlineRow";
import ConfirmDialog from "@/components/Modal/ConfirmDialog";

/**
 * RouteSettingsSection
 *
 * This component provides a settings section for managing a planned route.
 * Users can view, delete, and import routes while receiving guidance on
 * creating and uploading their own route files.
 *
 * @returns {React.ReactElement}
 *
 *
 */
const DeleteIcon = (props?: Partial<ImageProps>): IconElement => (
  <Icon {...props} name="trash-can" style={[props?.style, { height: 24 }]} />
);

const ImportIcon = (props?: Partial<ImageProps>): IconElement => {
  const theme = useTheme(); // Access the theme
  return (
    <Icon
      {...props}
      name="file-import"
      style={[props?.style, { height: 24 }]} // Size of the icon
      fill={theme["color-primary-600"]} // Set the color explicitly
    />
  );
};

type TourenParams = {
  tourId: string;
  tourTitle: string;
};

export default function RouteSettingsSection() {
  const params = useLocalSearchParams() as TourenParams;
  const { tourId } = params;
  const [selectedFile, setSelectedFile] =
    useState<DocumentPicker.DocumentPickerResult>();
  const [showDialog, setShowDialog] = useState(false);
  const [routeExists, setRouteExists] = useState(false);
  const theme = useTheme();
  const styles = makeStyles(theme);

  useEffect(() => {
    const fetchRoute = async () => {
      const route = await getTourRoute(tourId);
      if (route) {
        setRouteExists(true);
      }
    };

    void fetchRoute();
  }, [tourId]);

  const handleDeleteRoute = async () => {
    await deleteRoute(tourId);
    setRouteExists(false);
    setShowDialog(false);
  };

  const importTour = async () => {
    const file = await DocumentPicker.getDocumentAsync({
      type: "application/json",
    });
    if (!file.canceled) {
      const content = await FileSystem.readAsStringAsync(file.assets[0].uri);
      if (Gjv.valid(JSON.parse(content))) {
        setSelectedFile(file);
        await setTourRoute(tourId, content);
        setRouteExists(true);
      } else {
        Alert.alert("Fehler", "Die Datei ist kein gültiges GeoJSON");
      }
    }
  };

  return (
    <SingleSettingLayout settingName={"Geplante Route"}>
      <>
        <InlineRow
          leftComponent={<Text category={"h6"}>GeoJSON-Datei</Text>}
          buttons={
            <>
              {routeExists && (
                <TouchableOpacity
                  onPress={() => setShowDialog(true)}
                  testID={"delete-button"}
                >
                  <DeleteIcon style={styles.iconStyle} />
                </TouchableOpacity>
              )}
              <TouchableOpacity onPress={importTour} testID={"import-button"}>
                <ImportIcon style={styles.iconStyle} />
              </TouchableOpacity>
            </>
          }
        />
        <View>
          <Text style={styles.description}>
            Die geplante Route wird als graue Linie auf der Karte dargestellt
            und kann während dem Fahren zur Navigation benutzt werden. Falls du
            eine eigene Route erstellen möchtest, gehe wie folgt vor:
          </Text>
          <Text style={styles.description}>
            {"\u2022"} Besuche{" "}
            <Text
              style={[styles.accentText, { textDecorationLine: "underline" }]}
              onPress={() => {
                Linking.openURL(
                  "https://maps.openrouteservice.org/#/place/@11.832275390625002,48.622016428468406,7",
                ).catch((err) => console.error("Failed to open URL:", err));
              }}
            >
              {"OpenRouteService"}
            </Text>
            .
          </Text>
          <Text style={styles.description}>
            {"\u2022"} Setze Punkte auf der Karte, um deine Route zu planen.
            Beachte, dass das Planen der Route länger dauern kann, wenn die
            Punkte weiter auseinander liegen.
          </Text>
          <Text style={styles.description}>
            {"\u2022"} Exportiere die Route als Geo-JSON Datei auf dein Gerät.
          </Text>
          <Text style={styles.description}>
            {"\u2022"} Importiere die Datei hier, um sie zu verwenden.
          </Text>
        </View>
        {selectedFile?.assets?.at(0)?.name && (
          <Text style={styles.successMessage}>
            {selectedFile?.assets?.at(0)!.name} {"\n"} wurde erfolgreich
            importiert!
          </Text>
        )}
        <ConfirmDialog
          title="Route löschen"
          message="Möchtest du deine importierte Route wirklich löschen? Diese Aktion kann nicht rückgängig gemacht werden."
          confirmString="Löschen"
          cancelString="Abbrechen"
          visible={showDialog}
          onConfirm={handleDeleteRoute}
          onCancel={() => {
            setShowDialog(false);
          }}
        />
      </>
    </SingleSettingLayout>
  );
}

const makeStyles = (theme: ThemeType) =>
  StyleSheet.create({
    successMessage: {
      marginTop: 10,
      color: theme["color-primary-600"],
    },
    description: {
      marginTop: 5,
      color: theme["color-basic-500"],
    },
    accentText: {
      color: theme["color-primary-600"],
    },
    iconStyle: {
      height: 25,
      width: "auto",
      color: theme["color-primary-500"],
    },
  });
