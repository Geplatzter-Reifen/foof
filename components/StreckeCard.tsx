import React from 'react';
import {View, Text, StyleSheet} from 'react-native';
import {DATE_TIME, dateFormat} from "@/utils/dateFormatter";

export type StreckeData = {
  startLoc: string,
  endLoc: string,
  startTime: Date,
  endTime: Date,
}

export default function StreckeCard(props: StreckeData) {
  return (
    <View style={styles.card}>
      <Text style={styles.title}>Strecke</Text>
      <Text style={styles.label}>Start:</Text>
      <Text style={styles.info}>{props.startLoc}</Text>
      <Text style={styles.label}>Ziel:</Text>
      <Text style={styles.info}>{props.endLoc}</Text>
      <Text style={styles.label}>Startzeitpunkt:</Text>
      <Text style={styles.info}>{dateFormat(props.startTime, DATE_TIME)}</Text>
      <Text style={styles.label}>Endzeitpunkt:</Text>
      <Text style={styles.info}>{dateFormat(props.endTime, DATE_TIME)}</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  card: {
    backgroundColor: '#f5f5f5',
    width: '90%',
    padding: 20,
    marginVertical: 10,
    borderRadius: 10,
    shadowColor: '#000',
    shadowOffset: {width: 0, height: 2},
    shadowOpacity: 0.1,
    shadowRadius: 4,
    elevation: 2,
  },
  title: {
    fontSize: 18,
    fontWeight: 'bold',
    color: 'green',
    marginBottom: 10,
  },
  label: {
    fontSize: 14,
    fontWeight: 'bold',
    color: '#333',
    marginTop: 5,
  },
  info: {
    fontSize: 16,
    color: '#666',
  },
});