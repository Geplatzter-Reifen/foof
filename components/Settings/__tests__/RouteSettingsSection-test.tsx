import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import { ApplicationProvider } from "@ui-kitten/components";
import * as eva from "@eva-design/eva";
import RouteSettingsSection from "../RouteSettingsSection";
import { Alert } from "react-native";

jest.mock("expo-router", () => ({
  useLocalSearchParams: jest.fn(() => ({
    tourId: "mock-tourId",
    tourTitle: "Mock Title",
  })),
}));

jest.mock("@/services/data/routeService", () => ({
  deleteRoute: jest.fn(),
  getTourRoute: jest.fn().mockResolvedValue("mock-route-data"),
  setTourRoute: jest.fn(),
}));

jest.mock("expo-document-picker", () => ({
  getDocumentAsync: jest.fn(),
}));

jest.mock("expo-file-system", () => ({
  readAsStringAsync: jest.fn(),
}));

jest.mock("geojson-validation", () => ({
  valid: jest.fn(),
}));

jest.spyOn(Alert, "alert").mockImplementation(() => {});

afterEach(() => {
  jest.clearAllMocks();
  jest.useRealTimers();
});

jest.mock("@ui-kitten/components", () => {
  const originalModule = jest.requireActual("@ui-kitten/components");
  return {
    ...originalModule,
    Icon: jest.fn(({ name }) => `MockedIcon(${name})`), // Mock Icon rendering
  };
});

describe("RouteSettingsSection", () => {
  it("matches the snapshot on initial render", () => {
    const { toJSON } = render(
      <ApplicationProvider {...eva} theme={eva.light}>
        <RouteSettingsSection />
      </ApplicationProvider>,
    );

    expect(toJSON()).toMatchSnapshot(); // Save the snapshot for initial render
  });

  it("handles deleting a route by showing modal", async () => {
    const { getByTestId, findByText } = render(
      <ApplicationProvider {...eva} theme={eva.light}>
        <RouteSettingsSection />
      </ApplicationProvider>,
    );

    await waitFor(() => {
      expect(getByTestId("delete-button")).toBeTruthy();
    });
    fireEvent.press(getByTestId("delete-button"));
    expect(findByText("Route löschen")).toBeTruthy();
  });

  it("shows an error for invalid GeoJSON files", async () => {
    const mockFile = {
      canceled: false,
      assets: [{ name: "invalid.geojson", uri: "file://invalid.geojson" }],
    };

    require("expo-document-picker").getDocumentAsync.mockResolvedValue(
      mockFile,
    );
    require("expo-file-system").readAsStringAsync.mockResolvedValue(
      JSON.stringify({ type: "InvalidType" }),
    );
    require("geojson-validation").valid.mockReturnValue(false);

    const { getByTestId } = render(
      <ApplicationProvider {...eva} theme={eva.light}>
        <RouteSettingsSection />
      </ApplicationProvider>,
    );

    fireEvent.press(getByTestId("import-button"));

    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledWith(
        "Fehler",
        "Die Datei ist kein gültiges GeoJSON",
      );
    });
  });
});
