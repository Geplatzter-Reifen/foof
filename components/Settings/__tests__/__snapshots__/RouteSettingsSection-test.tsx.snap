// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteSettingsSection matches the snapshot on initial render 1`] = `
<View
  appearance="default"
  level="1"
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
      },
      {
        "marginBottom": 15,
      },
    ]
  }
>
  <Text
    appearance="default"
    category="s2"
    status="basic"
    style={
      [
        {
          "color": "#222B45",
          "fontFamily": "System",
          "fontSize": 13,
          "fontWeight": "600",
        },
        {
          "color": "#C5CEE0",
        },
      ]
    }
  >
    Geplante Route
  </Text>
  <View
    appearance="default"
    style={
      [
        {
          "backgroundColor": "#EDF1F7",
          "height": 1,
        },
        {
          "alignSelf": "stretch",
          "backgroundColor": "#E4E9F2",
          "height": 1,
          "marginVertical": 3,
        },
      ]
    }
  />
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 10,
      }
    }
  >
    <View
      style={
        {
          "width": "70%",
        }
      }
    >
      <Text
        appearance="default"
        category="h6"
        status="basic"
        style={
          [
            {
              "color": "#222B45",
              "fontFamily": "System",
              "fontSize": 18,
              "fontWeight": "800",
            },
            undefined,
          ]
        }
      >
        GeoJSON-Datei
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "gap": 10,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="import-button"
      >
        MockedIcon(file-import)
      </View>
    </View>
  </View>
  <View>
    <Text
      appearance="default"
      category="p1"
      status="basic"
      style={
        [
          {
            "color": "#222B45",
            "fontFamily": "System",
            "fontSize": 15,
            "fontWeight": "400",
          },
          {
            "color": "#C5CEE0",
            "marginTop": 5,
          },
        ]
      }
    >
      Die geplante Route wird als graue Linie auf der Karte dargestellt und kann während dem Fahren zur Navigation benutzt werden. Falls du eine eigene Route erstellen möchtest, gehe wie folgt vor:
    </Text>
    <Text
      appearance="default"
      category="p1"
      status="basic"
      style={
        [
          {
            "color": "#222B45",
            "fontFamily": "System",
            "fontSize": 15,
            "fontWeight": "400",
          },
          {
            "color": "#C5CEE0",
            "marginTop": 5,
          },
        ]
      }
    >
      •
       Besuche
       
      <Text
        appearance="default"
        category="p1"
        onPress={[Function]}
        status="basic"
        style={
          [
            {
              "color": "#222B45",
              "fontFamily": "System",
              "fontSize": 15,
              "fontWeight": "400",
            },
            [
              {
                "color": "#274BDB",
              },
              {
                "textDecorationLine": "underline",
              },
            ],
          ]
        }
      >
        OpenRouteService
      </Text>
      .
    </Text>
    <Text
      appearance="default"
      category="p1"
      status="basic"
      style={
        [
          {
            "color": "#222B45",
            "fontFamily": "System",
            "fontSize": 15,
            "fontWeight": "400",
          },
          {
            "color": "#C5CEE0",
            "marginTop": 5,
          },
        ]
      }
    >
      •
       Setze Punkte auf der Karte, um deine Route zu planen. Beachte, dass das Planen der Route länger dauern kann, wenn die Punkte weiter auseinander liegen.
    </Text>
    <Text
      appearance="default"
      category="p1"
      status="basic"
      style={
        [
          {
            "color": "#222B45",
            "fontFamily": "System",
            "fontSize": 15,
            "fontWeight": "400",
          },
          {
            "color": "#C5CEE0",
            "marginTop": 5,
          },
        ]
      }
    >
      •
       Exportiere die Route als Geo-JSON Datei auf dein Gerät.
    </Text>
    <Text
      appearance="default"
      category="p1"
      status="basic"
      style={
        [
          {
            "color": "#222B45",
            "fontFamily": "System",
            "fontSize": 15,
            "fontWeight": "400",
          },
          {
            "color": "#C5CEE0",
            "marginTop": 5,
          },
        ]
      }
    >
      •
       Importiere die Datei hier, um sie zu verwenden.
    </Text>
  </View>
</View>
`;
